polyfill/lib/ecmascript.mjs

```diff
@@ -310,6 +311,22 @@ export const ES = ObjectAssign(ObjectAssign({}, ES2019), {
     }
     return any ? any : false;
   },
+  ValidDateTimeFrom: (bag, required, optional = []) => {
+    if (!bag || 'object' !== typeof bag) return false;
+    let result = {};
+    for (let prop of required) {
+      if (!(prop in bag)) throw new TypeError(`required property '${prop}' missing`);
+      const value = ES.ToNumber(bag[prop]);
+      if (Number.isFinite(value)) result[prop] = value;
```

Comment:
This is a bit weird... doesn't it mean that e.g. `Temporal.DateTime.from({ month: 1, day: 1 })` throws an error while `Temporal.DateTime.from({ year: undefined, month: 1, day: 1 })` returns a Temporal.DateTime instance for 0000-01-01T00:00Z?

<details><summary><code>Temporal.DateTime.from({ year: undefined, month: 1, day: 1 })</code></summary>

1. [Temporal.DateTime.from(arg)](https://github.com/tc39/proposal-temporal/blob/8bbaf4c1569ceede161f17bd79d1670e3d20aca3/polyfill/lib/datetime.mjs#L298): `let result = ES.ToDateTime(arg)`
   1. [ES.ToDateTime(item)](https://github.com/tc39/proposal-temporal/blob/8bbaf4c1569ceede161f17bd79d1670e3d20aca3/polyfill/lib/ecmascript.mjs#L110): `const props = ES.ValidDateTimeFrom(item, ['year', 'month', 'day'], …)`
      1. [ES.ValidDateTimeFrom(bag, required, optional = [])](https://github.com/tc39/proposal-temporal/blob/8bbaf4c1569ceede161f17bd79d1670e3d20aca3/polyfill/lib/ecmascript.mjs#L319): `const value = ES.ToNumber(bag[prop])`
         1. [ToNumber(argument)](https://tc39.es/ecma262/#sec-tonumber): argument type undefined ⇒ return **NaN**
      1. [ES.ValidDateTimeFrom(bag, required, optional = [])](https://github.com/tc39/proposal-temporal/blob/8bbaf4c1569ceede161f17bd79d1670e3d20aca3/polyfill/lib/ecmascript.mjs#L320): `if (Number.isFinite(value)) result[prop] = value;` (_condition not satisfied for **NaN**, so property not assigned_)
   1. [ES.ToDateTime(item)](https://github.com/tc39/proposal-temporal/blob/8bbaf4c1569ceede161f17bd79d1670e3d20aca3/polyfill/lib/ecmascript.mjs#L123): `const { hour = 0, day, microsecond = 0, millisecond = 0, minute = 0, month, nanosecond = 0, second = 0, year, } = props` (_`year` is **undefined**_)
   1. [ES.ToDateTime(item)](https://github.com/tc39/proposal-temporal/blob/8bbaf4c1569ceede161f17bd79d1670e3d20aca3/polyfill/lib/ecmascript.mjs#L134): `return new TemporalDateTime(year, month, day, hour, minute, second, millisecond, microsecond, nanosecond, 'reject');`
      1. [Temporal.DateTime.prototype.constructor( year, month, day, … )](https://github.com/tc39/proposal-temporal/blob/8bbaf4c1569ceede161f17bd79d1670e3d20aca3/polyfill/lib/datetime.mjs#L33): `year = ES.ToInteger(year);`
         1. [ToInteger(argument)](https://tc39.es/ecma262/#sec-tointeger): Let <var>number</var> be ?&nbsp;<emu-xref aoid="ToNumber" id="_ref_1810"><a href="https://tc39.es/ecma262/#sec-tonumber">ToNumber</a></emu-xref>(<var>argument</var>).
            1. [ToNumber(argument)](https://tc39.es/ecma262/#sec-tonumber): argument type undefined ⇒ return **NaN**
         1. [ToInteger(argument)](https://tc39.es/ecma262/#sec-tointeger): If <var>number</var> is <emu-val>NaN</emu-val>, <emu-val>+0</emu-val>, or <emu-val>-0</emu-val>, return <emu-val>+0</emu-val>. (_condition satisfied for `year`_)
      1. [Temporal.DateTime.prototype.constructor( year, month, day, … )](https://github.com/tc39/proposal-temporal/blob/8bbaf4c1569ceede161f17bd79d1670e3d20aca3/polyfill/lib/datetime.mjs#L76): `SetSlot(this, YEAR, year);`
</details>

Why should there be a difference between implicit vs. explicit `undefined`?

Reply:
There should not; the spec generally uses `Get`, not `HasProperty`.

Reply:
Fixed this, and added tests where one item of the required properties for Date, DateTime, YearMonth, and MonthDay is undefined, ensuring that from() throws in that case.